[UnifiedPDF] [iPad] Unable to tap on first page for VoiceOver to read

ginalined opened the pull request:
#### c688e4779a63742a56e15410219c89ef9fa79e96
<pre>
[UnifiedPDF] [iPad] Unable to tap on first page for VoiceOver to read
<a href="https://bugs.webkit.org/show_bug.cgi?id=292385">https://bugs.webkit.org/show_bug.cgi?id=292385</a>
<a href="https://rdar.apple.com/148356538">rdar://148356538</a>

Reviewed by Abrar Rahman Protyasha.

- Fix the problem that sometimes the accessibilityElements are not fetched when conducting the hitTest
- Add support for accessibilityScroll

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::accessibilityCoreObject):
* Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
(-[WKAccessibilityPDFDocumentObject accessibilityHitTest:]):
(-[WKAccessibilityPDFDocumentObject accessibilityScroll:]):

Canonical link: <a href="https://commits.webkit.org/294457@main">https://commits.webkit.org/294457@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68fa4db791f568de551edfbf55aa50ab6ab6c910

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101836 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21504 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11820 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106994 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52470 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21812 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30011 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77537 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34560 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104843 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16849 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91956 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57875 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16676 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51821 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86529 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10051 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109351 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28969 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21334 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86515 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29330 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88157 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86087 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21913 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30847 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8566 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/23140 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28897 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34188 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28708 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32031 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30267 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3739,7 +3739,17 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+    if (!m_frame || !m_frame->page() || !m_frame->coreLocalFrame())
+        return nil;
+
+    if (WebCore::AXObjectCache* cache = UnifiedPDFPlugin::axObjectCache())
```
The prefix is not necessary.

```suggestion
    if (WebCore::AXObjectCache* cache = axObjectCache())
```

aprotyas replied:
I would suggest using a `CheckedPtr` since we expect the object cache to outlive this scope.
```suggestion
    if (CheckedPtr cache = axObjectCache())
```

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3739,7 +3739,17 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+    if (WebCore::AXObjectCache* cache = UnifiedPDFPlugin::axObjectCache())
+        return cache->getOrCreate(m_element.get());
+    return nil;
+}
```
Please return `nullptr` for C++ pointers

```suggestion
WebCore::AXCoreObject* UnifiedPDFPlugin::accessibilityCoreObject()
{
    if (!m_frame || !m_frame->page() || !m_frame->coreLocalFrame())
        return nullptr;

    if (WebCore::AXObjectCache* cache = UnifiedPDFPlugin::axObjectCache())
        return cache->getOrCreate(m_element.get());
    return nullptr;
}
```

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3739,7 +3739,17 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+
+WebCore::AXCoreObject* UnifiedPDFPlugin::accessibilityCoreObject()
+{
+    if (!m_frame || !m_frame->page() || !m_frame->coreLocalFrame())
```
Is this check necessary?

ginalined replied:
Right I move away from m_frame as I am working on it. I will delete it

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3739,7 +3739,17 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+        return cache->getOrCreate(m_element.get());
+    return nil;
+}
+#endif // !PLATFORM(IOS_FAMILY)
```
Typo
```suggestion
#endif // PLATFORM(IOS_FAMILY)
```

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
```diff
@@ -142,10 +144,13 @@ - (NSArray *)accessibilityElements
 }
-
-#endif // !PLATFORM(MAC)
+#endif // !PLATFORM(IOS_FAMILY)
```
Typo
```suggestion
#endif // PLATFORM(IOS_FAMILY)
```

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -30,6 +30,9 @@
 #include "PDFPluginBase.h"
+#if PLATFORM(IOS_FAMILY)
+#include <WebCore/AXCoreObject.h>
+#endif // PLATFORM(IOS_FAMILY)
```
You can get away with a forward declaration, so maybe shift the include from `UnifiedPDFPlugin.h` to `UnifiedPDFPlugin.mm`.

aprotyas replied:
The forward declaration would look like `class AXCoreObject;` under the `namespace WebCore` section in this header.

Commit:
[UnifiedPDF] [iPad] Unable to tap on first page for VoiceOver to read
https://bugs.webkit.org/show_bug.cgi?id=292385
rdar://148356538

Reviewed by Abrar Rahman Protyasha.

- Fix the problem that sometimes the accessibilityElements are not fetched when conducting the hitTest
- Add support for accessibilityScroll

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::accessibilityCoreObject):
* Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
(-[WKAccessibilityPDFDocumentObject accessibilityHitTest:]):
(-[WKAccessibilityPDFDocumentObject accessibilityScroll:]):

Canonical link: https://commits.webkit.org/294457@main
